build_mt.py

```diff
@@ -1,599 +1,827 @@
+
+        # print percentage where no information is available or halo not eligible
+        print(
+            "percentage no info or ineligible = ",
+            np.sum(mask_noinfo_this) / len(mask_noinfo_this) * 100.0,
+        )
+
+        # no info is denoted by 0 or -999 (or regular if ineligible), but -999 messes with unpacking, so we set it to 0
+        main_prog_this[mask_noinfo_this] = 0
+
+        # rework the main progenitor and halo indices to return in proper order
+        main_prog_this = correct_inds(
+            main_prog_this,
+            N_halos_slabs_prev,
+            slabs_prev,
+            start=start_prev,
+            stop=stop_prev,
+            copies=copies_prev,
+        )
+        halo_ind_this = correct_inds(
+            halo_ind_this,
+            N_halos_slabs_this,
+            slabs_this,
+            start=start_this,
+            stop=stop_this,
+            copies=copies_this,
+        )
+        halo_ind_prev = correct_inds(
+            halo_ind_prev,
+            N_halos_slabs_prev,
+            slabs_prev,
+            start=start_prev,
+            stop=stop_prev,
+            copies=copies_prev,
+        )
+
+        # we only use this when loading incomplete merger trees because we're missing data
+        if stop_this != n_chunks:
+            mask_noinfo_this[main_prog_this > N_halos_prev] = False
+            main_prog_this[main_prog_this > N_halos_prev] = 0
+
+        # positions and comoving distances of main progenitor halos corresponding to the halos in current snapshot
+        pos_prev_main_this = pos_prev[main_prog_this]
+        com_dist_prev_main_this = com_dist_prev[main_prog_this]
+        halo_ind_prev_main_this = halo_ind_prev[main_prog_this]
+
+        # halos that have merger tree information
+        pos_this_info = pos_this[mask_info_this]
+        com_dist_this_info = com_dist_this[mask_info_this]
+        halo_ind_this_info = halo_ind_this[mask_info_this]
+        pos_prev_main_this_info = pos_prev_main_this[mask_info_this]
+        com_dist_prev_main_this_info = com_dist_prev_main_this[mask_info_this]
+        halo_ind_prev_main_this_info = halo_ind_prev_main_this[mask_info_this]
+        eligibility_this_info = eligibility_this[mask_info_this]
+        if "Progenitors" in fields_mt:
+            start_progs_this_info = start_progs_this[mask_info_this]
+            num_progs_this_info = num_progs_this[mask_info_this]
+
+        # halos that don't have merger tree information
+        pos_this_noinfo = pos_this[mask_noinfo_this]
+        com_dist_this_noinfo = com_dist_this[mask_noinfo_this]
+        halo_ind_this_noinfo = halo_ind_this[mask_noinfo_this]
+        eligibility_this_noinfo = eligibility_this[mask_noinfo_this]
+        # pos_prev_main_this_noinfo = pos_prev_main_this[mask_noinfo_this]
+        # com_dist_prev_main_this_noinfo = com_dist_prev_main_this[mask_noinfo_this]
```

Comment:
Jumping ahead a bit, in wherever the algorithm lands, we'll probably want to use Astropy Tables to organize these "parallel" arrays.  That should collapse these repeated lines into one line.

Reply:
yep, yep, totally agree!